src/libstore/build/local-derivation-goal.cc

```diff
@@ -2323,6 +2327,9 @@ DrvOutputs LocalDerivationGoal::registerOutputs()
             }, *orifu);
         }},
         {[&](const std::string & path, const std::string & parent) {
+            // as the result is cast to an Error, the handler for BuildErrors is not run
```

Comment:
Sounds like this is the root cause.
I think we could either make the exception type a parameter of `topoSort`.

Shifting the throw to the callback would preserve the simple type signature, but `[[noreturn]] void()` in a template produces a warning in gcc. Maybe just `void ()` is good enough? It doesn't encode that the callback must throw, but it gets the job done without warning fatigue.

I don't have a strong opinion on the details, but I think this should be solved by allowing `topoSort` to throw arbitrary exceptions.

Reply:
I tried to make it a template argument however the compiler couldn't infer the type. I could restrict it to `buildError`s as that is all it is used with, however I will have the closure throw the error.